File: go/vectorize/ir_vec.go
package hyper

import (
	"fmt"
	"github.com/llir/llvm/ir"
	"github.com/llir/llvm/ir/constant"
	"github.com/llir/llvm/ir/enum"
	"github.com/llir/llvm/ir/types"
	"github.com/llir/llvm/ir/value"
)

func isFloat(opcode enum.BinaryOp) bool {
	switch opcode {
	case enum.FAdd, enum.FSub, enum.FMul, enum.FDiv, enum.FRem:
		return true
	default:
		return false
	}
}

type BinaryIROperation struct {
	Opcode   enum.BinaryOp
	Bitwidth int
}

func (op *BinaryIROperation) Match(v value.Value, matches *[]Match) bool {
	if binOp, ok := v.(*ir.InstBinOp); ok {
		if binOp.Op == op.Opcode && hasBitWidth(v, op.Bitwidth) {
			*matches = append(*matches, Match{
				LiveIn:   false,
				Operands: []value.Value{binOp.X, binOp.Y},
				Result:   v,
			})
			return true
		}
	}
	return false
}

func (op *BinaryIROperation) GetMaximumVF(tti *TargetTransformInfo) int {
	return tti.GetLoadStoreVecRegBitWidth(0) / op.Bitwidth
}

func (op *BinaryIROperation) GetName() string {
	return fmt.Sprintf("%s-i%d", op.Opcode, op.Bitwidth)
}

type UnaryIROperation struct {
	Opcode   enum.UnaryOp
	Bitwidth int
}

func (op *UnaryIROperation) Match(v value.Value, matches *[]Match) bool {
	if unaryOp, ok := v.(*ir.InstUnaryOp); ok {
		if unaryOp.Op == op.Opcode && hasBitWidth(v, op.Bitwidth) {
			*matches = append(*matches, Match{
				LiveIn:   false,
				Operands: []value.Value{unaryOp.X},
				Result:   v,
			})
			return true
		}
	}
	return false
}

func (op *UnaryIROperation) GetMaximumVF(tti *TargetTransformInfo) int {
	return tti.GetLoadStoreVecRegBitWidth(0) / op.Bitwidth
}

func (op *UnaryIROperation) GetName() string {
	return fmt.Sprintf("%s-%d", op.Opcode, op.Bitwidth)
}

type IRVectorBinding struct {
	Op      *BinaryIROperation
	Name    string
	Sig     InstSignature
	LaneOps []BoundOperation
}

func (b *IRVectorBinding) GetCost(tti *TargetTransformInfo, ctx *types.Context) float64 {
	var scalarTy types.Type
	if isFloat(b.Op.Opcode) {
		if b.Op.Bitwidth == 32 {
			scalarTy = types.Float
		} else {
			scalarTy = types.Double
		}
	} else {
		scalarTy = types.NewInt(b.Op.Bitwidth)
	}
	numElems := len(b.LaneOps)
	vecTy := types.NewVector(scalarTy, int64(numElems))
	return tti.GetArithmeticInstrCost(b.Op.Opcode, vecTy)
}

func CreateIRVectorBinding(op *BinaryIROperation, vectorWidth int) *IRVectorBinding {
	sig := InstSignature{
		InputWidths:  []int{vectorWidth, vectorWidth},
		OutputWidths: []int{vectorWidth},
		HasImm8:      false,
	}
	elemWidth := op.Bitwidth
	numLanes := vectorWidth / elemWidth
	laneOps := make([]BoundOperation, numLanes)
	for i := 0; i < numLanes; i++ {
		lo := i * elemWidth
		hi := lo + elemWidth
		laneOps[i] = BoundOperation{
			Op: op,
			InputBindings: []InputBinding{
				{0, lo, hi},
				{1, lo, hi},
			},
		}
	}
	return &IRVectorBinding{
		Op:      op,
		Name:    op.GetName(),
		Sig:     sig,
		LaneOps: laneOps,
	}
}

func (b *IRVectorBinding) Emit(operands []value.Value, builder *ir.Builder) (value.Value, error) {
	if len(operands) != 2 {
		return nil, fmt.Errorf("invalid number of operands")
	}
	return builder.NewBinOp(b.Op.Opcode, operands[0], operands[1], "")
}

func (b *IRVectorBinding) IsSupported(tti *TargetTransformInfo) bool {
	return len(b.LaneOps) <= b.Op.GetMaximumVF(tti)
}

type UnaryIRVectorBinding struct {
	Op      *UnaryIROperation
	Name    string
	Sig     InstSignature
	LaneOps []BoundOperation
}

func (b *UnaryIRVectorBinding) GetCost(tti *TargetTransformInfo, ctx *types.Context) float64 {
	elemWidth := b.Op.Bitwidth
	numElems := len(b.LaneOps)
	if b.Op.Opcode == enum.FNeg {
		var ty types.Type
		if elemWidth == 32 {
			ty = types.Float
		} else {
			ty = types.Double
		}
		vecTy := types.NewVector(ty, int64(numElems))
		return tti.GetArithmeticInstrCost(enum.FNeg, vecTy)
	}
	return 1.0
}

func CreateUnaryIRVectorBinding(op *UnaryIROperation, vectorWidth int) *UnaryIRVectorBinding {
	sig := InstSignature{
		InputWidths:  []int{vectorWidth},
		OutputWidths: []int{vectorWidth},
		HasImm8:      false,
	}
	elemWidth := op.Bitwidth
	numLanes := vectorWidth / elemWidth
	laneOps := make([]BoundOperation, numLanes)
	for i := 0; i < numLanes; i++ {
		lo := i * elemWidth
		hi := lo + elemWidth
		laneOps[i] = BoundOperation{
			Op: op,
			InputBindings: []InputBinding{
				{0, lo, hi},
			},
		}
	}
	return &UnaryIRVectorBinding{
		Op:      op,
		Name:    op.GetName(),
		Sig:     sig,
		LaneOps: laneOps,
	}
}

func (b *UnaryIRVectorBinding) Emit(operands []value.Value, builder *ir.Builder) (value.Value, error) {
	if len(operands) != 1 {
		return nil, fmt.Errorf("invalid number of operands")
	}
	ctx := builder.GetInsertBlock().Parent.Parent.Context()
	elemWidth := b.Op.Bitwidth
	numElems := len(b.LaneOps)
	floatTy := types.Float
	if elemWidth == 64 {
		floatTy = types.Double
	}
	intTy := types.NewInt(elemWidth)
	vecFloatTy := types.NewVector(floatTy, int64(numElems))
	vecIntTy := types.NewVector(intTy, int64(numElems))
	v := operands[0]
	switch b.Op.Opcode {
	case enum.SIToFP:
		return builder.NewSIToFP(tryCast(builder, v, vecIntTy), vecFloatTy)
	case enum.FPToSI:
		return builder.NewFPToSI(tryCast(builder, v, vecFloatTy), vecIntTy)
	case enum.FNeg:
		return builder.NewFNeg(tryCast(builder, v, vecFloatTy))
	default:
		return nil, fmt.Errorf("unsupported unary opcode")
	}
}

func (b *UnaryIRVectorBinding) IsSupported(tti *TargetTransformInfo) bool {
	return len(b.LaneOps) <= b.Op.GetMaximumVF(tti)
}

func tryCast(builder *ir.Builder, v value.Value, ty types.Type) value.Value {
	if v.Type() == ty {
		return v
	}
	return builder.NewBitCast(v, ty)
}

File: ocaml/src/vectorize/ir_vec.ml
open Llvm
open Llvm_analysis
open Llvm_target

let is_float = function
  | Fadd | Fsub | Fmul | Fdiv | Frem -> true
  | _ -> false

module BinaryIROperation = struct
  type t = {
    opcode: Opcode.t;
    bitwidth: int;
  }

  let match op v matches =
    match value_kind v with
    | Instruction (Binop bin_op) when bin_op = op.opcode && has_bit_width v op.bitwidth ->
      let operands = [| get_operand v 0; get_operand v 1 |] in
      matches := { live_in = false; operands; result = v } :: !matches;
      true
    | _ -> false

  let get_maximum_vf op tti =
    load_store_vec_reg_bit_width tti 0 / op.bitwidth

  let get_name op =
    Printf.sprintf "%s-i%d" (string_of_opcode op.opcode) op.bitwidth
end

module UnaryIROperation = struct
  type t = {
    opcode: Opcode.t;
    bitwidth: int;
  }

  let match op v matches =
    match value_kind v with
    | Instruction i when instruction_opcode i = op.opcode && has_bit_width v op.bitwidth ->
      let operands = [| get_operand v 0 |] in
      matches := { live_in = false; operands; result = v } :: !matches;
      true
    | _ -> false

  let get_maximum_vf op tti =
    load_store_vec_reg_bit_width tti 0 / op.bitwidth

  let get_name op =
    Printf.sprintf "%s-%d" (string_of_opcode op.opcode) op.bitwidth
end

module IRVectorBinding = struct
  type t = {
    op: BinaryIROperation.t;
    name: string;
    sig_: InstSignature.t;
    lane_ops: BoundOperation.t list;
  }

  let get_cost b tti ctx =
    let scalar_ty =
      if is_float b.op.opcode then
        if b.op.bitwidth = 32 then float_type ctx
        else double_type ctx
      else integer_type ctx b.op.bitwidth
    in
    let num_elems = List.length b.lane_ops in
    let vec_ty = vector_type scalar_ty num_elems in
    arithmetic_instr_cost tti b.op.opcode vec_ty

  let create op vector_width =
    let sig_ = {
      input_widths = [vector_width; vector_width];
      output_widths = [vector_width];
      has_imm8 = false;
    } in
    let elem_width = op.bitwidth in
    let num_lanes = vector_width / elem_width in
    let lane_ops = List.init num_lanes (fun i ->
      let lo = i * elem_width in
      let hi = lo + elem_width in
      BoundOperation.create op [|(0, lo, hi); (1, lo, hi)|]
    ) in
    { op; name = BinaryIROperation.get_name op; sig_; lane_ops }

  let emit b operands builder =
    match operands with
    | [|op1; op2|] -> build_binop b.op.opcode op1 op2 "" builder
    | _ -> failwith "Invalid number of operands"

  let is_supported b tti =
    List.length b.lane_ops <= BinaryIROperation.get_maximum_vf b.op tti
end

module UnaryIRVectorBinding = struct
  type t = {
    op: UnaryIROperation.t;
    name: string;
    sig_: InstSignature.t;
    lane_ops: BoundOperation.t list;
  }

  let get_cost b tti ctx =
    let elem_width = b.op.bitwidth in
    let num_elems = List.length b.lane_ops in
    match b.op.opcode with
    | Fneg ->
      let ty = if elem_width = 32 then float_type ctx else double_type ctx in
      let vec_ty = vector_type ty num_elems in
      arithmetic_instr_cost tti b.op.opcode vec_ty
    | _ -> 1.0

  let create op vector_width =
    let sig_ = {
      input_widths = [vector_width];
      output_widths = [vector_width];
      has_imm8 = false;
    } in
    let elem_width = op.bitwidth in
    let num_lanes = vector_width / elem_width in
    let lane_ops = List.init num_lanes (fun i ->
      let lo = i * elem_width in
      let hi = lo + elem_width in
      BoundOperation.create op [|(0, lo, hi)|]
    ) in
    { op; name = UnaryIROperation.get_name op; sig_; lane_ops }

  let emit b operands builder =
    match operands with
    | [|op|] ->
      let ctx = context builder in
      let elem_width = b.op.bitwidth in
      let num_elems = List.length b.lane_ops in
      let float_ty = if elem_width = 32 then float_type ctx else double_type ctx in
      let int_ty = integer_type ctx elem_width in
      let vec_float_ty = vector_type float_ty num_elems in
      let vec_int_ty = vector_type int_ty num_elems in
      let v = op in
      match b.op.opcode with
      | Sitofp -> build_sitofp (try_cast builder v vec_int_ty) vec_float_ty "" builder
      | Fptosi -> build_fptosi (try_cast builder v vec_float_ty) vec_int_ty "" builder
      | Fneg -> build_fneg (try_cast builder v vec_float_ty) "" builder
      | _ -> failwith "Unsupported unary opcode"
    | _ -> failwith "Invalid number of operands"

  let is_supported b tti =
    List.length b.lane_ops <= UnaryIROperation.get_maximum_vf b.op tti
end